### PR TITLE
Add CPU Pinning document

### PR DESCRIPTION
**Pull request by brandboat**
related to https://github.com/harvester/harvester/issues/6148, add cpu pinning document.

### CODE REVIEW

File: docs/vm/cpu-pinning.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+2. Click the hamburger button.
+3. Select either Enable CPU Manager or Disable CPU Manager.
+4. Wait ~1 minute for the action to complete.
+
+![enable-cpu-manager](/img/v1.4/cpu-pinning/enable-cpu-manager.png)
+![disable-cpu-manager](/img/v1.4/cpu-pinning/disable-cpu-manager.png)
+
+### Limitations
+
+- Only one master node can enable/disable the CPU Manager at a time.
```
Comment (ibrokethecloud) [inline docs/vm/cpu-pinning.md:64]: we need to rephrase this, since the CPU manager can be enable on any node in the cluster except the witness node

Comment (brandboat) [reply docs/vm/cpu-pinning.md]: Thanks, already refine the limitations part, also add description about what harvester behaves after user checks the `enable CPU pinning` checkbox in virtual machine page. As you mentioned offline, this would help user understand how to enable cpu pinning via YAML.

File: docs/vm/cpu-pinning.md [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+2. Go to Virtual Machines page, click Create. 
+3. Click Advanced Options.
+![create-vm](/img/v1.4/cpu-pinning/create-vm.png)
+4. Scroll down to the bottom, check the Enable CPU Pinning checkbox.
+![vm-advanced-options](/img/v1.4/cpu-pinning/vm-advanced-options.png)
+
+Enabling the checkbox adds `dedicatedCpuPlacement: true` to `.spec.template.spec.domain.cpu` in the virtual machine YAML. In Harvester, when we detect that `dedicatedCpuPlacement` is set to true in the virtual machine definition, we automatically set the CPU and memory resource requests equal to the limits to ensure it meets the criteria for Guaranteed QoS.
+
+- To enable CPU pinning for an existing VM, restart the VM after enabling this option to apply the change.
+- CPU and memory [Resource Overcommit](./resource-overcommit.md) does not apply to VMs with CPU pinning, as they use Guaranteed QoS, meaning their requests and limits are the same.
```
Comment (jillian-maroket) [inline docs/vm/cpu-pinning.md:80]:
```suggestion
Enabling CPU pinning adds `dedicatedCpuPlacement: true` to `.spec.template.spec.domain.cpu` in the virtual machine configuration (YAML). When `dedicatedCpuPlacement` is set to `true`, the CPU and memory resource requests are automatically set to match the limits to ensure that the criteria for Guaranteed QoS are met.

Because the requests and limits are identical, CPU and memory [Resource Overcommit](./resource-overcommit.md) settings do not apply to VMs with CPU pinning enabled.

:::note

To use CPU pinning on an existing VM, you must restart the VM after enabling the feature and saving the change.

:::
```

File: docs/vm/cpu-pinning.md [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+## VM Live Migration
+The migration of a VM with CPU pinning is mentioned in [Live Migration](./live-migration.md). However, you must ensure that the target node has CPU manager enabled; otherwise, the migration will not succeed.
+
+## Upgrade
+During an upgrade, Harvester will drain all pods on the node and live migrate VMs to another node. If there are VMs with CPU pinning enabled, ensure that the other nodes have the CPU Manager enabled. Otherwise, the upgrade process may get stuck.
```
Comment (jillian-maroket) [inline docs/vm/cpu-pinning.md:86]:
```suggestion
## VM Live Migration

VMs with CPU pinning enabled can be migrated only if the CPU Manager is enabled on the target node. For more information, see [Live Migration](./live-migration.md).

## Upgrades

When upgrading a node, Harvester drains all pods and live migrates VMs to another node. To avoid interruptions to the upgrade process, ensure that the CPU Manager is enabled on other nodes whenever you use VMs with CPU pinning enabled.
```

**Review by jillian-maroket — COMMENTED** [review]
Let me know if you have concerns about the changes.

File: docs/vm/cpu-pinning.md [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+4. Wait ~1 minute for the action to complete.
+
+![enable-cpu-manager](/img/v1.4/cpu-pinning/enable-cpu-manager.png)
+![disable-cpu-manager](/img/v1.4/cpu-pinning/disable-cpu-manager.png)
+
+### Limitations
+
+- Only one master node can enable or disable the CPU Manager at a time. Note that the witness node cannot perform these actions.
+- You must wait for an operation (enable/disable) to finish before starting another on the same node.
+- Any VMs with CPU pinning enabled must be stopped before disabling CPU Manager on that node.
```
[the remark this answers is not in the document]
Comment (ibrokethecloud) [reply docs/vm/cpu-pinning.md:66]: CPU manager is not limited to management/controlplane node, it can be enable on any node in the cluster. We shoudl rephrase this to better reflect this

`You can enable or disable the CPU Manager on only one management node at a time.`

[the remark this answers is not in the document]
Comment (brandboat) [reply docs/vm/cpu-pinning.md:66]: Thanks for the comment @ibrokethecloud ! What I want to emphasize here is that users could only enable/disable cpu manager  on one management node **_at the same time_.** 

Just a note, I add this limitation since enable/disable cpu manager requires restarting rke2-server on management node, if all rke2-server are restarting in the same time, that would cause whole cluster unavailable.

File: docs/vm/cpu-pinning.md (hunk 3 above)
Comment (brandboat) [reply docs/vm/cpu-pinning.md:86]: I changed the sentence a bit:

When upgrading a node, Harvester drains all pods and live migrates VMs to another node. To avoid interruptions to the upgrade process, ensure that the CPU Manager is enabled on other nodes and sufficient resources available whenever you use VMs with CPU pinning enabled.

Only added `and sufficient resources available` here.

**Review by a110605 — APPROVED** [review]
The doc looks clear to me.

File: docs/vm/cpu-pinning.md [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+Allow some time for Harvester to apply the corresponding CPU Manager policy.
+
+![enable-cpu-manager](/img/v1.4/cpu-pinning/enable-cpu-manager.png)
+![disable-cpu-manager](/img/v1.4/cpu-pinning/disable-cpu-manager.png)
+
+### Limitations
+
+- The CPU Manager cannot be enabled on the witness node.
+
+- You can enable or disable the CPU Manager on only one management node at a time.
```
Comment (bk201) [inline docs/vm/cpu-pinning.md:70]: This sounds like when a user enables CPU manager on a management node, he/she can't enable another management node. It would be nice to re-phase it more clearly.

File: docs/vm/cpu-pinning.md [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+1. Click **Create**.
+
+  ![create-vm](/img/v1.4/cpu-pinning/create-vm.png)
+
+1. On the **Advanced Options** tab, select **Enable CPU Pinning**.
+
+  ![vm-advanced-options](/img/v1.4/cpu-pinning/vm-advanced-options.png)
+
+1. On the **Advanced Options** tab, select **Enable CPU Pinning**.
```
Comment (jillian-maroket) [inline docs/vm/cpu-pinning.md:90]: The step was somehow duplicated. Please remove these lines.

File: docs/vm/cpu-pinning.md [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+
+:::
+
+## VM Live Migration
+
+VMs with CPU pinning enabled can be migrated only if the CPU Manager is enabled on the target node. For more information, see [Live Migration](./live-migration.md).
+
+## Upgrades
+
+When upgrading a node, Harvester drains all pods and live migrates VMs to another node. To avoid interruptions to the upgrade process, ensure that the CPU Manager is enabled on other nodes and sufficient resources available whenever you use VMs with CPU pinning enabled.
```
Comment (jillian-maroket) [inline docs/vm/cpu-pinning.md:112]:
```suggestion
When upgrading a node, Harvester drains all pods and live migrates VMs to another node. To avoid interruptions to the upgrade process, ensure that the CPU Manager is enabled on other nodes and sufficient resources are available whenever you use VMs with CPU pinning enabled.
```

File: docs/vm/cpu-pinning.md [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+![enable-cpu-manager](/img/v1.4/cpu-pinning/enable-cpu-manager.png)
+![disable-cpu-manager](/img/v1.4/cpu-pinning/disable-cpu-manager.png)
+
+### Limitations
+
+- The CPU Manager cannot be enabled on the witness node.
+
+- You can enable or disable the CPU Manager on only one management node at a time.
+
+- You must wait for an operation (enabling or disabling) to be completed before starting another.
```
Comment (jillian-maroket) [inline docs/vm/cpu-pinning.md:72]:
```suggestion
- The CPU Manager must be enabled or disabled on each node separately. You must wait for the operation to be completed before starting another.
```

Comment (jillian-maroket) [reply docs/vm/cpu-pinning.md:72]: @brandboat Let's combine points 2 and 3 to address @bk201 's comment.

**Review by jillian-maroket — COMMENTED** [review]
I responded to the feedback and recently added/modified text.

File: docs/vm/cpu-pinning.md [hunk 9]
```diff
@@ -105,8 +99,8 @@ To use CPU pinning on an existing VM, you must restart the VM after enabling the
 
 ## VM Live Migration
 
-VMs with CPU pinning enabled can be migrated only if the CPU Manager is enabled on the target node. For more information, see [Live Migration](./live-migration.md).
+VMs with CPU pinning enabled can be migrated only if the CPU Manager is enabled on the target node.
```
Comment (brandboat) [inline docs/vm/cpu-pinning.md:102]: I removed the reference to `For more information, see [Live Migration](./live-migration.md).` to avoid misleading users into thinking we provide additional information about live migration with CPU manager enabled, which we currently do not.

**Review by jillian-maroket — APPROVED** [review]
LGTM. Thanks!

**Review by bk201 — APPROVED** [review]
LGTM and thanks for rephrasing.
